### ISSUE TITLE
Remove pkgsettings

### ISSUE THREAD

**Issue by marcosschroh** [kpn/kstreams#6]
Right now we are using `pkgsettings` to read `env` variables and then initialize `consumers` and `producer`.

```python
class Consumer(aiokafka.AIOKafkaConsumer):
    def __init__(
        self,
        *args,
        settings_prefix: str = "SERVICE_KSTREAMS_",
        key_deserializer: Optional[Callable] = None,
        **kwargs,
    ):
        self.settings = PrefixedSettings(conf.settings, prefix=settings_prefix)

        if key_deserializer is None:
            key_deserializer = lambda k: None if k is None else k.decode()  # noqa: E731

        kafka_config = utils.retrieve_kafka_config(settings_prefix=settings_prefix)
        kafka_config.update(kwargs)

        super().__init__(*args, key_deserializer=key_deserializer, **kafka_config)
``` 

we want to remove `pkgsettings` and then use maybe `pydantic` to read `settings` and the engine will receive all the values:

```python
stream_engine = create_engine(
    title="my-stream-engine",
    security_protocol="...",
    ssl_context="...",
    ...
    value_serializer=JsonSerializer(),
    value_deserializer=JsonDeserializer(),
)
```

**Comment by Woile** [comment]
I have different proposals for this.

One thing for sure that would be useful, is to have a kafka settings class, so it can be passed around and reused (multiple instances of `create_engine`, using `create_engine` + a `Consumer`, etc.

Now the question is how to use that class, see the examples:

---

# A. Use a `KafkaSettings` class as parameter

```python
@dataclass
class KafkaSettings:
    bootstrap_servers: List[str]
    security_protocol: SecurityProtocol = SecurityProtocol.PLAINTEXT
    ssl_cabundle_data: Optional[str] = None
    ssl_cert_data: Optional[str] = None
    ssl_key_data: Optional[str] = None

    def ssl_context(self) -> Optional[ssl.SSLContext]:
         .... 
```

```
settings = KafkaSettings(...)
stream_engine = create_engine(
   settings=settings
)
```

## B. Unpack the settings 

```python
@dataclass
class KafkaSettings:
    bootstrap_servers: List[str]
    security_protocol: SecurityProtocol = SecurityProtocol.PLAINTEXT
    ssl_cabundle_data: Optional[str] = None
    ssl_cert_data: Optional[str] = None
    ssl_key_data: Optional[str] = None

   def asdict(self):
           return dataclasses.asdict(self)
```

```
settings = KafkaSettings(...)
stream_engine = create_engine(
   **settings.asdict()
)
```

**Comment by marcosschroh** [comment]
Good question. We need to solve 2 problems:

1. How to read the settings:
2. How to use the settings with the engine.

**Read the settings:**

A. We can use `dataclasses` like you mentioned 
B. We can use `pydantic`
C. Do nothing. End users will have to take care of how to read the settings and validate them.

I think having a `KafkaSettings` will help. It is reusable, it can validate properties, etc.

**How to use the settings with the engine.** 

I think being explicit is a good idea (`Unpack the settings`) because you will have the 2 options to load the settings into the engine: either using the `KafkaSettings` or don't (option C for users that don't want anything or they have a custom settings read process). It will have a lot of flexibility.

```python
# Use KafkaSettings 
settings = KafkaSettings(...)

stream_engine = create_engine(
   **settings.asdict()
) 

# or

stream_engine = create_engine(
   ...
   # settings
   security_protocol="...",
   ssl_context="...",
    ...
)
```

In addition to this, we could have something like `Flask` does: a `engine.load_settings(settings)` method, which will merge the settings and override previous values. S

**Comment by Woile** [comment]
After discussing with marcos. We agreed on using `stream_engine = create_engine(settings=settings)`. This way settings makes it easy to change backends.

**Comment by Woile** [comment]
I'm exploring a different approach: I'm going to use a `TypedDict` to validate the dictionary entries:

For example, given these settings:

```python
class KafkaPlaintextSettings(TypedDict):
    bootstrap_servers: List[str]
    security_protocol: Literal[SecurityProtocol.PLAINTEXT]


class KafkaSslSettings(TypedDict):
    bootstrap_servers: List[str]
    security_protocol: Literal[SecurityProtocol.SSL]
    ssl_context: ssl.SSLContext
```


We would validate against:

```python
def create_engine(..., setttings: Union[KafkaPlaintextSettings, KafkaSslSettings]):
    ...
```

This way, doing something like:

```python
stream_engine = create_engine(setttings={"bootstrap_servers": ["localhost:9092"], security_protocol: "plaintext"})
```

Would not be linted by mypy, but doing:

```python
stream_engine = create_engine(setttings={"bootstrap_servers": ["localhost:9092"]})
```

Would show a mypy error.

This removes the need to import and instanciate a class, while still having the typing benefits.

Thoughts?

**Comment by marcosschroh** [comment]
Cool, I like the idea.

**Comment by Woile** [comment]
Update on the typeddict proposal, mypy is not smart enough to figure out that a dict matches a typed dict.

This:
```python
stream_engine = create_engine(
    setttings={"bootstrap_servers": ["localhost:9092"], "security_protocol": "plaintext"}
)
```

fails with this error:
```
Argument "settings" to "Producer" has incompatible type "Dict[str, Sequence[str]]"; expected "Union[KafkaPlaintextConfig, KafkaSslConfig, None]
```

This means that we have to look for a different approach.

Atm I'm considering the following:

```
broker = brokers.Kafka(bootstrap_servers=[])
stream_engine = create_engine(title="my-stream-engine", broker=broker)
```

**Comment by Woile** [comment]
Closed by #35
